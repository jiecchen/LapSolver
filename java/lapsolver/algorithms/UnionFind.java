/**
 * @file UnionFind.java
 * @author Daniel Spielman <spielman@cs.yale.edu>
 * @date Fri May 30 2014
 *
 * A crude Union Find data structure, implemented using linked lists for each set.
 * Should eventually be replaced by a pointer-based structure with path compression.
 *
 */

package lapsolver.algorithms;

public class UnionFind {
    //-----------------------
    //   the number of verts
    //
    public int nv;

    //-----------------------
    //
    // comp[i] is the name of the component of vertex i
    public int[] comp;

    //-----------------------
    //
    // ptr[i] is a pointer to the next element in the same set as item i
    // -1 if it is the end of a list
    public int[] ptr;

    //-----------------------
    //
    // lastPtr[i] is the last item in comp[i], 
    // so ptr[lastPtr[i]] = -1;
    public int[] lastPtr;


    //-----------------------
    //
    public int[] compSize;

    // Call with number of items in the sets
    public UnionFind(int n) {
        this.nv = n;

        comp = new int[n];
        ptr = new int[n];
        lastPtr = new int[n];
        compSize = new int[n];

        for (int i = 0; i < n; i++) {
            ptr[i] = -1;
            compSize[i] = 1;
            comp[i] = i;
            lastPtr[i] = i;
        }
    }

    public int find(int v) {
        return comp[v];
    }

    // merge the components with items u and v
    public void union(int u, int v) {
        int uc = comp[u];
        int vc = comp[v];

        // swap to make vc the smaller one
        if (compSize[vc] > compSize[uc]) {
            int tmp = vc;
            vc = uc;
            uc = tmp;
        }

        // go through the list for vc, re-naming comp for all
        int i = vc;
        while (i != -1) {
            comp[i] = uc;
            i = ptr[i];
        }

        compSize[uc] += compSize[vc];
        compSize[vc] = 0;

        ptr[lastPtr[uc]] = vc;
        lastPtr[uc] = lastPtr[vc];
    }
}
