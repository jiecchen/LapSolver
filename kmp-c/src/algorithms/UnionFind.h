#pragma once

class UnionFind {
  public:
    UnionFind(int n);
    ~UnionFind();
    void link(int u, int v);
    int find_set(int u);

  private:
    int* parent;
    int* to_compress;
};
