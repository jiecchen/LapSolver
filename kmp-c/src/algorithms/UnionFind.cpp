#include "UnionFind.h"

UnionFind::UnionFind(int n)
{
    to_compress = new int[n];
    parent = new int[n];
    for (int i = 0; i < n; i++)
        parent[i] = i;
}

UnionFind::~UnionFind()
{
    delete[] parent;
}

void UnionFind::link(int u, int v)
{
    int pu = find_set(u), pv = find_set(v);
    if (pu != pv) parent[pu] = pv;
}

int UnionFind::find_set(int u)
{
    int pos = 0;

    // follow pointers until self-loop
    while (parent[u] != u)
    {
        to_compress[pos++] = u;
        u = parent[u];
    }

    // Path compression
    parent[to_compress[0:pos]] = u;
    return u;
}
