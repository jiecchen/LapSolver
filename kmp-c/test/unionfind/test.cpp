// Test for UnionFind

#include <cstdlib>
#include <iostream>
#include <string>
#include <algorithms/UnionFind.h>

using namespace std;

int main(int argc, char *argv[])
{
    // read n and number of links
    int n, q;
    cin >> n >> q;

    UnionFind uf(n);

    // perform links
    for (int i = 0; i <= q; i++) {
        int u, v;
        cin >> u >> v;
        uf.link(u, v);
        
        // report state after each link
        for (int j = 0; j < n; j++) {
            cout << uf.find_set(j) << ' ';
        } cout << '\n';
    }

    return 0;
}
