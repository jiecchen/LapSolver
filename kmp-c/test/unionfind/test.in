8
9
5 5
0 1
2 3
4 5
6 7
0 3
4 7
7 4
1 6