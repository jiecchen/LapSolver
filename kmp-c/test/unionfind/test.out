0 1 2 3 4 5 6 7 
1 1 2 3 4 5 6 7 
1 1 3 3 4 5 6 7 
1 1 3 3 5 5 6 7 
1 1 3 3 5 5 7 7 
3 3 3 3 5 5 7 7 
3 3 3 3 7 7 7 7 
3 3 3 3 7 7 7 7 
7 7 7 7 7 7 7 7 
7 7 7 7 7 7 7 7 
